### PR TITLE
ref(scheduler): add HTTP functions in KubeHTTPClient

### LINKED ISSUE #1008: gracefully handle cases when the k8s api server refuses connections
As seen in internal support calls with GKE, it is entirely possible for the controller/database pod to be operating normally while the k8s API server refuses to accept incoming connections. We need to handle cases during this time so that we can recover gracefully from these situations without breaking application deploys.

### PR DESCRIPTION
**Pull request by bacongobbler**
These functions properly handle connection errors and clean up some of the child/parent class
relations when making a request.

closes #1008

### CODE REVIEW

**Comment by helgi** [comment]
Awesome, wonder if we should handle the `ConnectTimeout` and `ReadTimeout` exceptions `requests` has? There is also a `SSLError` one but that's perhaps a step too far.

Btw this can go into 2.5 since it was delayed.

File: rootfs/scheduler/tests/test_kubehttpclient.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        """
+        self.adapter.register_uri('GET', self.url)
+        response = self.scheduler.http_get(self.path)
+        assert response is not None
+        self.assertTrue(self.adapter.called)
+        self.assertEqual(self.adapter.call_count, 1)
+        # ensure that connection errors get raised as a KubeException
+        self.adapter.add_matcher(connection_refused_matcher)
+        with self.assertRaises(exceptions.KubeException):
+            self.scheduler.http_head(self.path)
```
Comment (helgi) [inline rootfs/scheduler/tests/test_kubehttpclient.py]: Should this be `http_head` when it is in `test_get`?

File: rootfs/scheduler/tests/test_kubehttpclient.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        """
+        self.adapter.register_uri('POST', self.url)
+        response = self.scheduler.http_post(self.path, data=json.dumps({'hello': 'world'}))
+        assert response is not None
+        self.assertTrue(self.adapter.called)
+        self.assertEqual(self.adapter.call_count, 1)
+        # ensure that connection errors get raised as a KubeException
+        self.adapter.add_matcher(connection_refused_matcher)
+        with self.assertRaises(exceptions.KubeException):
+            self.scheduler.http_head(self.path)
```
Comment (helgi) [inline rootfs/scheduler/tests/test_kubehttpclient.py]: Should this be `http_head` when it is in `test_post`?

File: rootfs/scheduler/tests/test_kubehttpclient.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        """
+        self.adapter.register_uri('PUT', self.url)
+        response = self.scheduler.http_put(self.path, data=json.dumps({'hello': 'world'}))
+        assert response is not None
+        self.assertTrue(self.adapter.called)
+        self.assertEqual(self.adapter.call_count, 1)
+        # ensure that connection errors get raised as a KubeException
+        self.adapter.add_matcher(connection_refused_matcher)
+        with self.assertRaises(exceptions.KubeException):
+            self.scheduler.http_head(self.path)
```
Comment (helgi) [inline rootfs/scheduler/tests/test_kubehttpclient.py]: Should this be `http_head` when it is in `test_put`?

File: rootfs/scheduler/tests/test_kubehttpclient.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        """
+        self.adapter.register_uri('DELETE', self.url)
+        response = self.scheduler.http_delete(self.path)
+        assert response is not None
+        self.assertTrue(self.adapter.called)
+        self.assertEqual(self.adapter.call_count, 1)
+        # ensure that connection errors get raised as a KubeException
+        self.adapter.add_matcher(connection_refused_matcher)
+        with self.assertRaises(exceptions.KubeException):
+            self.scheduler.http_head(self.path)
```
Comment (helgi) [inline rootfs/scheduler/tests/test_kubehttpclient.py]: Should this be `http_head` when it is in `test_delete`?

File: rootfs/scheduler/tests/test_kubehttpclient.py (hunk 1 above)
Comment (bacongobbler) [inline rootfs/scheduler/tests/test_kubehttpclient.py]: good catch. will fix

Comment (bacongobbler) [inline rootfs/scheduler/tests/test_kubehttpclient.py]: all fixed.

**Pull request merged**